operator/v1/types_ingress.go

```diff
@@ -1399,6 +1399,132 @@ type IngressControllerHTTPHeaders struct {
+	// and spec.httpHeaders.headerNameCaseAdjustments.
+	// The header settings made here can be overwritten by spec.httpHeaders or annotations
+	// on Route objects.
+	// Setting httpCaptureHeaders for the headers set via this API will not work.
+	// Setting "cookie" or "set-cookie" headers that conflict with OpenShift router's session cookies will result in undefined behavior.
+	// Note that the total size of all added headers *after* interpolating dynamic values
+	// must not exceed the value of spec.tuningOptions.headerBufferMaxRewriteBytes on the
+	// IngressController.
+	// +optional
+	Actions IngressControllerHTTPHeaderActions `json:"actions,omitempty"`
+}
+
+// IngressControllerHTTPHeaderActions defines configuration for actions on HTTP request and response headers.
+type IngressControllerHTTPHeaderActions struct {
+	// response is a list of HTTP response headers to set or delete.
+	// The cluster administrator uses these to specify HTTP response headers that should be set or
+	// deleted at the global level when forwarding responses from an application to the client, i.e. for all routes.
+	// The actions either Set or Delete will be performed on the headers in sequence as defined in the list of response headers.
+	// A maximum of 20 response header actions may be configured.
+	// Sample fetchers allowed are "res.hdr" and "ssl_c_der".
+	// Converters allowed are "lower" and "base64".
+	// Example header values: "%[res.hdr(X-target),lower]", "%{+Q}[ssl_c_der,base64]".
+	// + ---
+	// + Note: Any change to regex mentioned below must be reflected in the CRD validation of route in https://github.com/openshift/library-go/blob/master/pkg/route/validation/validation.go and vice-versa.
+	// +listType=map
+	// +listMapKey=name
+	// +optional
+	// +kubebuilder:validation:MaxItems=20
+	// +kubebuilder:validation:XValidation:rule=`self.all(key, key.action.type == "Delete" || (has(key.action.set) && key.action.set.value.matches('^(?:%(?:%|(?:\\{[-+]?[QXE](?:,[-+]?[QXE])*\\})?\\[(?:res\\.hdr\\([0-9A-Za-z-]+\\)|ssl_c_der)(?:,(?:lower|base64))*\\])|[^%[:cntrl:]])+$')))`,message="Either the header value provided is not in correct format or the sample fetcher/converter specified is not allowed. The dynamic header value will be interpreted as an HAProxy format string as defined in https://cbonte.github.io/haproxy-dconv/2.4/configuration.html#8.2.4 and may use HAProxy's %[] syntax and otherwise must be a valid HTTP header value as defined in https://datatracker.ietf.org/doc/html/rfc7230#section-3.2. Sample fetchers allowed are res.hdr, ssl_c_der. Converters allowed are lower, base64."
+	Response []IngressControllerHTTPHeader `json:"response"`
+	// request is a list of HTTP request headers to set or delete.
+	// The cluster administrator uses these to specify HTTP request headers that should be set or
+	// deleted at the global level when forwarding requests to your application from the client, i.e. for all routes.
```

Comment:
```suggestion
	// Actions defined here will modify the request headers of all requests passing through an ingress controller. These actions are applied to all Routes defined within a cluster.
	// Actions may be overwritten by actions defined locally on individual Route objects.
```

Reply:
I really don't like `The cluster administrator uses these to specify HTTP request headers that should be set or deleted at the global level`, it doesn't flow well. What part of the suggestion I made do you think is not capturing the intent of the original?

When writing the godoc for APIs, there's a certain tone that we aim for. With that, typically shorter more direct sentences are better. I believe I captured the sentiment of your suggestion but in the tone we aim for for the API documentation with my suggestion

Reply:
@JoelSpeed thats fine I can add your suggestion. The original ones gives better info like setting at global level which is more imp for users. However please let me ASAP if this does not sound good to you as I need to make the changes ASAP. Thanks!

Reply:
I have suggested to Miheer a couple times that we describe the interaction in terms of the order in which actions are applied.  For example:

> In case of HTTP request headers, the actions specified in `spec.httpHeaders.actions` on the Route will be executed after the actions specified in the IngressController's `spec.httpHeaders.actions` field.

And conversely:

> In case of HTTP response headers, the actions specified in `spec.httpHeaders.actions` on the IngressController will be executed after the actions specified in the Route's `spec.httpHeaders.actions` field.

I believe that this language is more accurate and will make it clear, for example, that an action on a Route could use the value of a header that was set by an IngressController action, or an IngressController action could delete a response header that was set by a Route action.  This order-of-execution language also makes it clear that all the actions are applied, even if a Route action and an IngressController action both specify the same request or response header; the "override" language is less clear IMO.  @JoelSpeed, what do you think?

Reply:
Yeah that makes a lot of sense, can we update to use Miciah's order of applied language on this one then?

Reply:
> I have suggested to Miheer a couple times that we describe the interaction in terms of the order in which actions are applied. For example:
> 

I had already applied this on Actions 
	// Header actions defined for request here will not take precedence over any header modifying action or route annotation defined on Route objects.
	// Header actions defined for response here will take precedence over any header modifying action or route annotation defined on Route objects.

> > In case of HTTP request headers, the actions specified in `spec.httpHeaders.actions` on the Route will be executed after the actions specified in the IngressController's `spec.httpHeaders.actions` field.
> 
> And conversely:
> 
> > In case of HTTP response headers, the actions specified in `spec.httpHeaders.actions` on the IngressController will be executed after the actions specified in the Route's `spec.httpHeaders.actions` field.
> 

So do you want me to add 
`//In case of HTTP request headers, the actions specified in spec.httpHeaders.actions on the Route will be executed after the actions specified in the IngressController's spec.httpHeaders.actions field.`  above `Request []IngressControllerHTTPHeader `json:"request"``

AND

also replace 	`// Header actions defined for request here will not take precedence over any header modifying action or route annotation defined on Route objects.` with `//In case of HTTP request headers, the actions specified in spec.httpHeaders.actions on the Route will be executed after the actions specified in the IngressController's spec.httpHeaders.actions field.` ?


> I believe that this language is more accurate and will make it clear, for example, that an action on a Route could use the value of a header that was set by an IngressController action, or an IngressController action could delete a response header that was set by a Route action. This order-of-execution language also makes it clear that all the actions are applied, even if a Route action and an IngressController action both specify the same request or response header; the "override" language is less clear IMO. @JoelSpeed, what do you think?


Also here we are discussing about this `	// Actions defined here will modify the response headers of all requests passing through an ingress controller.
	// These actions are applied to all Routes i.e. for all connections handled by the ingress controller defined within a cluster.`

Right ?

This does not mention about the execution order but just an info about what routes will will pass through ingress controller. So from my understanding we will need to add this above Request [] ?
	// Actions defined here will modify the response headers of all requests passing through an ingress controller.
	// These actions are applied to all Routes i.e. for all connections handled by the ingress controller defined within a cluster.
      //  In case of HTTP request headers, the actions specified in spec.httpHeaders.actions on the Route will be executed after the actions specified in the IngressController's spec.httpHeaders.actions field.

Reply:
Yes I think we might be in the wrong place. At the point you mention the ordering, lets try and keep it short and to the point, something like (for the Route):
```
// Actions defined here will be executed after the actions specified in the IngressController's `spec.httpHeaders.actions` field.
```
And then the contrary for the IngressController version, have I got that the correct way around?

Reply:
In terms of ordering, you need to break it down into actions for request headers and actions for response headers: 
* IngressController actions for _request_ headers will be executed _before_ Route actions.  
* IngressController actions for _response_ headers will be executed _after_ Route actions.